http3/client.go

```diff
@@ -208,6 +264,36 @@ func (c *SingleDestinationRoundTripper) roundTrip(req *http.Request) (*http.Resp
 		return nil, err
 	}
 
+	id := str.StreamID()
+	ctx, cancel := context.WithCancelCause(req.Context())
+	c.ctxLock.Lock()
+	c.runningCtx[id] = cancel
+	c.ctxLock.Unlock()
+
+	defer func() {
+		cancel(nil)
+		c.ctxLock.Lock()
+		delete(c.runningCtx, id)
+		c.ctxLock.Unlock()
+	}()
+
+	// the context may not be canceled if it's added after goaway frame is processed
+	if goawayID, ok := c.goawayReceived(); ok && goawayID <= id {
+		cancel(errGoaway)
+		str.CancelWrite(quic.StreamErrorCode(ErrCodeRequestCanceled))
+		str.CancelRead(quic.StreamErrorCode(ErrCodeRequestCanceled))
+		return nil, errGoaway
+	}
+
+	// Separate goroutine to prevent interference with request cancellation
+	go func() {
```

Comment:
I might be missing something, but this looks to me like you're canceling requests that are in flight when receiving the GOAWAY. But that's exactly what we should not do: Requests still in flight should complete as if nothing happened.